### PR TITLE
First pass at updating validation language

### PR DESCRIPTION
**Pull request by jandrieu**
This is an attempt to revisit validation vs verification #1232

***
<a href="https://pr-preview.s3.amazonaws.com/LegReq/vc-data-model/pull/1297.html" title="Last updated on Oct 31, 2023, 4:43 PM UTC (400bcd8)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/vc-data-model/1297/286c2b2...LegReq:400bcd8.html" title="Last updated on Oct 31, 2023, 4:43 PM UTC (400bcd8)">Diff</a>

### CODE REVIEW

File: index.html
Comment (andresuribe87) [inline index.html:4655]: There is already a section called `Validation`. Right now it's appendix A. Changing this line will update all the existing anchor tags of the form `<a href="#validation">`, and they'll start pointing to this section, instead of the appendix. I've noted what you'd need to do to update one such example below.

File: index.html
Comment (OR13) [inline index.html:308]: how do we test this MUST?

File: index.html
Comment (OR13) [inline index.html:2520]: what is the difference between a "well-formed" and a "conforming document".

File: index.html
Comment (decentralgabe) [inline index.html:2385]: I like this!

File: index.html
Comment (decentralgabe) [inline index.html:5688]:
```suggestion
credential</a> is not in the past. Because some credentials can be useful for secondary purposes even if their original validity period has expired, validity period, as expressed using the <code>validFrom</code> and <code>validUntil</code> properties, is always considered a component of validation, which is performed **after** verification.
```

File: terms.html
Comment (decentralgabe) [inline terms.html:135]:
```suggestion
  verify <a>verifiable credentials</a> and <a>verifiable presentations</a>. 
```

File: terms.html
Comment (decentralgabe) [inline terms.html:142]:
```suggestion
  are outside the scope of this specification. It is expected that 
  <a>verifiers</a> will trust certain <a>issuers</a> for certain claims and apply their 
  own rules to determine which claims in which <a>credentials</a> are suitable
  for use by their systems.
```

**Review by decentralgabe — APPROVED** [review]
great changes - approving noting @andresuribe87 and my comments

File: index.html
Comment (OR13) [inline index.html:2555]: replace with issue marker, or text.

File: index.html
Comment (OR13) [reply index.html:4655]: I don't think its wise to make validation normative... it drastically increases test complexity, and it will trample on use cases.

File: index.html
Comment (OR13) [inline index.html:5688]: This is good, but see my comment above, which is related to this sentence here: 

 https://github.com/w3c/vc-data-model/pull/1297/files#r1342943590

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2023-10-11-vcwg#section1-8) on 2023-10-11

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 1.8. First pass at updating validation language (pr vc-data-model#1297)

_See github pull request [vc-data-model#1297](https://github.com/w3c/vc-data-model/pull/1297)._




**Brent Zundel:** Joe has a response to an open issue.  
… This updates the language in the spec around validation.  
… There has been conversation, suggested changes that aren't in yet.  
… There's at least one person that has approved it, we perhaps need more review.  

**Dave Longley:** I haven't had enough time to look at this one, initial scan seems to add normative requirements around statements that are not testable... Verifiers must validate claims, that's usually done via business rules... validate using included business rules... we'll need to adjust that language in some way... not something an implementation can do... you might need to use your own business rules.  
… Appreciate the effort to get clarity on verification vs. validation.  

**Manu Sporny:** The only concerning bits of the PR are around normative statements around validation. I know Jeffrey Yaskin wanted us to specify verification algorithms, but validation is a bridge too far.  
… There are multiple statements in this PR that say things like "verifiers MUST validate claims in VCs" which, again, is not testable. We don't want to get into validation rules.  

**Brent Zundel:** I agree with that. I wanted to ask if the normative text here might be ... one of the suggestions made was that we more clearly define conformance rules. We could say "a conformant consumer of a VC... ought to do", whether and how that language would be normative around that is a another discussion.  
… Perhaps putting this into conformant producers and consumers of the document is a step we could take. "Here's this data model and if you want to do something with it, here are some rules around that". It might be helpful if we went there.  

**Manu Sporny:** One approach we could take there is the approach we took in the DID spec for DID method spec authors. We had normative statements around what DID specs must do -- I think Jeffrey Yaskin's review also covered that there are conformance classes in the spec that we don't define like conforming issuer/verifier.  

> *Brent Zundel:* +1 that's what I was getting at.

**Manu Sporny:** I think that's what you were getting at Brent, so we could have sections in the spec about what a conforming verifier must or should do and hand wave over that and say -- while those statements are normative, we're just saying what's expected from the conformance class, now how to do it but that one must do it.  
… That might allow us to write normative statements and that we don't have entries for that in the test suite because it's general guidance.  

**Dave Longley:** it's "business rules", not something a generic implementation would do.  

**Manu Sporny:** Saying it out loud sounds problematic ... that's somewhere to try and go though.  

**Brent Zundel:** For this PR advice would be to get rid of normative language, recognizing that moving forward with conformance on issuer/verifier roles we might need some "more normative" language and that might not be testable, but separate PR for that.  

> *Manu Sporny:* +1 to approach brentz suggested.

</details>

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2023-10-17-vcwg-special#section2-7) on 2023-10-17

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 2.7. First pass at updating validation language (pr vc-data-model#1297)

_See github pull request [vc-data-model#1297](https://github.com/w3c/vc-data-model/pull/1297)._




**Brent Zundel:** Has an approval, quite a few suggestions, some conflicts that need resolving, I believe in general this PR is in good shape.  
… I think the last time we discussed it, the group landed that the validation section is good but for this PR ... it may be best to remove the normative text.  
… As we move forward, addressing the need for defining a conformative issuer or verifier, then normative language would fit better there. This is a reiteration of the conversation from last week, we just need a bit of clean up from Joe.  
… Joe is here -- Joe, would love to hear from you on this.  

**Joe Andrieu:** I think this is looking pretty good, need to deal with nuances from remaining comments.  

**Brent Zundel:** Are you ok with the path of removing the normative statements from this PR to address them more completely in a PR about conformance classes?  

**Joe Andrieu:** I guess I'm not sure what that means. I guess the normative statement is whether or not you must validate. Is that it?  

**Brent Zundel:** That's the primary example, but any normative language in this PR ... MUST/SHOULD, people have expressed concern over.  

**Joe Andrieu:** Let me read through it, I'll consider that. I need to understand the consequences before I commit.  

**Brent Zundel:** Thank you, Joe, I think this PR has a good chance for finding consensus, that was the sticking point though, ... one side requires normative language to be removed, so if you demand something else then it will be closed.  

**Joe Andrieu:** Ok, then I will take the action to remove it, and if I think we still need it, I will create another PR.  

**Brent Zundel:** Ok, thank you.  

**Manu Sporny:** Yeah, I was going to speak to the normative language, the concern is what Orie raised, if we have a MUST in there we will have to test in some way unless it's guidance to a spec writer or something. The easiest way here is just no RFC MUST language, that's the easiest thing to get in.  
… If you suggest in another PR to do some normative statements, if you're talking about validation, we'll have to define algorithms to do that. We have requests from Jeffrey Yaskin to create a normative algorithm for verification and one for validation is much more dicey.  
… Normative statements around validation will become stuff we have to test / have a good testing story around those statements.  

**Joe Andrieu:** I think we're in a pickle where testability might not be the right metric. Companies, what business rules they apply, etc. is out of scope -- that they do validate is a MUST. Before you use a property you must validate -- I get the issues here.  

**Brent Zundel:** I believe if you add the normative statements in the context of a conformance class -- the DID WG did something similar when we said "DID method authors MUST" ... and we don't have a way to test that and we still had normative requirements for that. I think that's a possible way forward.  
… If that makes sense at all.  

**Joe Andrieu:** That helps, I'll fold it into what I put together.  

</details>

File: index.html
Comment (jandrieu) [inline index.html:4656]:
```suggestion
      <section class="informative">
```

File: index.html
Comment (jandrieu) [reply index.html:4657]: @andresuribe87 I switched back to informative, as requested in yesterday's call. 

However, I'm not sure how to handle your suggestion. It seemed like you made it normative, which added the link problem. Does this problem still exist if this is non-normative?

File: index.html
Comment (jandrieu) [reply index.html:308]: It is not testable, nor does it need to be, imo.

How validation occurs is out of scope, but validating properties before using them is the only way this stuff works. @OR13 are you suggesting that verifiers use properties they don't understand?

Comment (OR13) [reply index.html:308]: I agree with the language, I was mostly commenting because every MUST changes the scope of the test suites, and in this case, I think we would decline to test this one, because as you said, its not testable as written.... its essentially saying "You MUST... do some undefined thing ... before you rely on claims".

File: index.html
Comment (OR13) [reply index.html:2520]: excellent, I defined the term here: https://github.com/w3c/vc-data-model/pull/1320

We use well formed in a few other places, so it will be helpful to define the term once.

Comment (jandrieu) [reply index.html:2520]: But that's not the definition that applies to this use.

Well-formed does not mean conformant to the VCDM. Well-formed in this usage means the shape of the data matches that specified by the value of the credentialSchema property.

File: index.html
Comment (jandrieu) [reply index.html:4657]: Except that verifiers are free to use their own rules and disregard validFrom and validUntil in cases where expiry doesn't matter.

The requirement isn't that verifiers must accept a VC as valid when the current timestamp is between validFrom and validUntil. The requirement is that verifiers must understand whether or not the validFrom or validUntil properties matter to their use case, and how.  That requirement is true for all properties a verifier relies on.

Once they understand their own desired policy, then they can build their validation rules to enforce that policy. How they come up with that policy and how it is enforced is not the W3C's business. It is literally the verifier's business.

File: index.html
Comment (jandrieu) [reply index.html:308]: While it is our commitment to test all testable normative requirements, this is a normative requirement, IMO, that is not testable, as all of the activity is downstream into and out of validation.

How validation is invoked is out of scope.
How validation is performed is out of scope.
How validation is reported is out of scope.
How verifiers handle the possible responses from validation is out of scope.

However, validation MUST occur before reliance on the property.

Validation is required to ensure the semantic layer of VCs, especially whether or not the issuer is an acceptable authority for a given claim and whether or not the cryptography is acceptable.

I think we would all be in agreement that, as a matter of principle, a verifier must not rely on properties for which they do not recognize the issuer as authoritative. 

That is the process of validation and, IMO, required. 

If its not normatively required, then we would be suggesting that verifiers can rely on properties just because they are signed and current--regardless of the issuer. That suggestion is clearly out of sync with the majority of use cases we have documented (where subject==holder) and basically the entire three-party model. One doesn't simply trust a VC because it verifiers, one must also make sure it is valid for the use case at hand. That is, verifiers MUST satisfy their own validation rules before relying on anything in a VC.

Given that fundamental aspect of VCs--that verifiers ALWAYS have to validate--it's clearly a MUST, despite the slight dissatisfaction from our inability to test that.

Comment (msporny) [inline index.html:308]:
```suggestion
that truth of <a>claims</a> encoded therein. Rather, once the authenticity and currency of a <a>verifiable credential</a> or <a>verifiable presentation</a> are established, a <a>verifier</a> validates the included claims using their own business rules before relying on them. Such reliance only occurs after evaluating the issuer, the proof, the subject, and the claims, against one or more verifier policies. 
```

Comment (msporny) [reply index.html:308]: The language is more clear without the RFC language. You just assert what you want to happen as fact, no need for RFC language here as it muddies the water from a testing perspective. We have to explain every normative statement that we're not testing, and this one would be hard to defend (and doesn't really add anything beyond just stating that verifiers validate stuff before they use it -- the ones that don't are bad verifiers and will get themselves into hot water (by not doing any validation) almost immediately).

File: index.html
Comment (msporny) [inline index.html:993]: The actual check is the `id` of the `credentialSubject` is associated, in some way, with the creator of the verifiable presentation. The language here is a bit loose, but that might be ok. I don't have a concrete suggestion that would be easy to read here.

File: index.html
Comment (msporny) [inline index.html:2557]: Please raise an issue for this and point to the issue from this issue marker.

File: index.html
Comment (msporny) [inline index.html:2643]:
```suggestion
  presentation</a>. However, <a>verifiers</a> 
  apply their own business rules before treating any claims as valid,
  taking into account the <a>holder</a>, the <a>issuers</a> of the
  <a>verifiable credentials</a>, the claims of the
  <a>verifiable credentials</a>, and its own policies.
```

File: index.html
Comment (msporny) [inline index.html:4664]:
```suggestion
When processing <a>verifiable credentials</a>, <a>verifiers</a>  
evaluate any relevant <a>claims</a> before relying upon them. This 
evaluation may be done in any manner desired, as long as it satisfies
the requirements of the <a>verifier</a> doing the validation. 
Many verifiers will perform the checks listed in Appendix <a href="#validation-0"></a> as well as a variety of specific business process checks such as:
```

File: index.html
Comment (msporny) [inline index.html:5690]:
```suggestion
credential</a> is not in the past. Because some credentials can be useful for secondary purposes even if their original validity period has expired, validity period, as expressed using the <code>validFrom</code> and <code>validUntil</code> properties, is always considered a component of validation, which is performed <em>after</em> verification.
```

File: terms.html
Comment (msporny) [inline terms.html:56]:
```suggestion
abstract concepts, numbers, and strings. Any entity might perform roles
```

File: terms.html
Comment (msporny) [inline terms.html:142]:
```suggestion
  It also specifies that <a>verifiers</a> validate claims in <a>verifiable credentials</a> 
  before relying on them. However, the means for such validation vary widely and 
  are outside the scope of this specification. It is expected that 
  <a>verifiers</a> will trust certain <a>issuers</a> for certain claims and apply their 
  own rules to determine which claims in which <a>credentials</a> are suitable
  for use by their systems.
```

**Review by msporny — CHANGES_REQUESTED** [review]
Overall, the explanatory text is good. I disagree that the untestable normative statements add value. I've suggested alternatives to the normative statements that have the impact that I believe you are seeking w/o potentially confusing implementers and test suite authors, or creating added burden on the Editors and staff to justify why the group is not normatively testing certain MUST statements.

File: index.html
Comment (jandrieu) [reply index.html]: I added an issue #1328 to track the normative nature of validation.

File: index.html
Comment (OR13) [reply index.html:4657]: you can resolve any thread with me, you feel is stale.

File: index.html
Comment (msporny) [inline index.html:4662]:
```suggestion
evaluation might be done in any manner desired, as long as it satisfies
```

File: terms.html
Comment (msporny) [inline terms.html:56]: While it's true that "numbers and strings" are examples of entities... I wonder if this is too meta for most readers of the document. I was trying to think of a "string" as a credential subject" and was struggling to think of anything other than a dictionary or thesaurus that would use that concept... and it's even harder to imagine a VC issued for a string or a number. 

Please consider removing the two examples and focusing more on entities that you'd expect to see a VC issued for (rather than more rare cases).

File: index.html
Comment (msporny) [reply index.html:2557]: Raised in issue #1331.

**Review by msporny — APPROVED** [review]
One very minor nit and one suggestion on terminology. Take the former, take it or leave it on the latter. This PR needs to be rebased.

**Review by longpd — APPROVED** [review]

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2023-10-31-vcwg-special#section1-8) on 2023-10-31

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 1.8. First pass at updating validation language (pr vc-data-model#1297)

_See github pull request [vc-data-model#1297](https://github.com/w3c/vc-data-model/pull/1297)._




**Brent Zundel:** Thank you, Manu, any other comments?  
… First pass at updating validation language. The last time we discussed this issue, there was agreement that everything was good except the normative text.  
… I'm not sure where things have gone since then.  

**Manu Sporny:** I believe Joe removed the normative text, it looks great at this point, I think it's ready to merge. Joe, I found one minor nit, and it needs a rebase.  
… Then an issue raised ... but this PR is ready to go with rebasing, etc.  

**Joe Andrieu:** Ok, I'm ok with that -- what issue?  

**Manu Sporny:** You created new text -- ...  

**Joe Andrieu:** Oh, I created that issue.  

**Manu Sporny:** Oh, maybe a duplicate there, I'll close my issue. This is ready to go, you just need to rebase.  

**Joe Andrieu:** Ok, I'll look at it. I agree we should be ready to merge.  

**Brent Zundel:** As soon as you rebase, we plan on merging.  
… As long as the comments have gotten in for Ivan's scripts.  
… Any other comments?  

</details>

File: index.html
Comment (jandrieu) [reply index.html:993]: "associated, in some way" is not correct.

This particular example is an example of redeeming an alumni discount from a university, in which case arbitrary association is not the business rule. Subject==Holder is.

**Comment by brentzundel** [comment]
@msporny this PR still cannot be rebased and merged, and it is unclear to me what the scope of work is to accomplish that. 
It could be squashed and merged. I understand that in general our preference is to rebase first, but in this case I think we should squash and merge so that we can continue moving forward.

**Comment by msporny** [comment]
> @msporny this PR still cannot be rebased and merged, and it is unclear to me what the scope of work is to accomplish that. It could be squashed and merged. I understand that in general our preference is to rebase first, but in this case I think we should squash and merge so that we can continue moving forward.

Something has gone horribly wrong with the rebase/update of this PR. I'll look into what it might take to fix it. A squash merge might be all we can do at this point.

> (Edit: I just re-reviewed this PR and see that now there are a number of "changes" that seem to be pulled in from other PRs. I wonder if a new clean PR with only the text changes ultimately agreed to here would be the simpler route forward.)

Yes, looks like Joe's main branch wasn't updated when he did an update/rebase, which caused all sorts of havoc in this branch... Joe updating his main branch and then rebasing this branch on top of the new main might fix the issue. I can try to do this manually when I get around to my next edit cycle. 

In any case, I'll endeavor to get this text into the spec as cleanly as possible (unless @jandrieu beats me to the task).

**Comment by iherman** [comment]
@msporny , the git(hub) wizzard :-)

**Comment by jandrieu** [comment]
FWIW, I'll try to get to that this weekend.

Some people are emotionally scarred from bad experiences in their youth. Rebase is like that for me.

**Comment by msporny** [comment]
Merged via rebased PR #1297, closing.

**Comment by msporny** [comment]
> Some people are emotionally scarred from bad experiences in their youth. Rebase is like that for me.

Well, that rebase was exceedingly painful. It took an hour to re-align everything. I don't know what happened in this PR, but let's please never do that again. :)

@jandrieu, you will want to check PR #1297 and make sure I didn't miss anything (I almost certainly did due to all the merge conflicts I had to resolve). If I did miss something, please raise a new PR to make the adjustments.

**Comment by TallTed** [comment]
> all the merge conflicts

For future reference, I've found that PRs with major merge conflicts are *far* more easily addressed in [BBEdit](http://bbedit.com/) than just using GitHub's tools. There are some manual steps, because you have to create 2 documents in BBEdit, one containing the "new" text and the other containing the "old", by manually editing all the `<<<<<<` and `>>>>>>` sections — but once that's done, it can show you character-level differences which you can apply in either direction, each with a single click. I work this way until the two documents are identical, and the whole thing gets pasted into GitHub atop the conflict display.

This has saved me *major* time on documents of 10,000+ lines with varyingly *huge* length, that couldn't have been resolved on GitHub alone without days of work, if ever. It's only available for macOS, but there may be some similar tools out there for Windows or Linux.

**Pull request closed without merging**
